wordpress-mobile/WordPress-iOS
# Blogging Prompts: Add handler for prompt notifications

Resolves #18739

This PR adds handling for both prompt notifications and static notifications. I intended to break this down into separate PRs, but it ended up looking manageable 😅 . Here are some notes regarding the changes:

- Updated notification payload to properly include `type`. This is required by `PushNotificationManager` if we want to add custom handling for notifications.
- Added `BloggingPromptCoordinator` that handles presenting the Post Creation flow.
  - We could reuse this component from the dashboard, FAB, and the feature introduction.
  - Also added a `Source` enum that's specific for Blogging Prompts. This is converted to `PostEditorEntryPoint` when showing the post creation flow. I was thinking this source enum could be useful for tracking later on.
- Edited some stuff in `WPTabBarController` so `PushNotificationManager` can trigger the post creation flow for the prompt without leaking too many implementation details.

## To Test

### Setup

To simplify testing in the simulator, here are some test codes to apply in `DashboardPromptsCardCell` and `WPTabBarController+Swift`:
<details>
<summary>Test code for <code>DashboardPromptsCardCell</code>:</summary>

```swift
// DashboardPromptsCardCell.swift

    func skipMenuTapped() {
//        saveSkippedPromptForSite()
//        presenterViewController?.reloadCardsLocally()


        guard let prompt = prompt,
              let siteID = blog?.dotComID?.intValue else {
            return
        }

        // show notification.
        let content = UNMutableNotificationContent()
        content.title = "Today's Prompts 💡"
        content.body = prompt.textForDisplay()
        content.userInfo = [
            PushNotificationsManager.Notification.typeKey: PushNotificationsManager.Notification.bloggingPrompts,
            PushNotificationsManager.BloggingPromptPayload.siteIDKey: siteID,
            PushNotificationsManager.BloggingPromptPayload.promptIDKey: Int(prompt.promptID)
        ]

        print(">>> Scheduling prompt with ID: \(prompt.promptID) for siteID: \(siteID)...")

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: TimeInterval(5), repeats: false)
        let request = UNNotificationRequest(identifier: "foo", content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request) { error in
            print("Error: \(error)")
        }
    }
```
</details>

<details>
<summary>Test code for <code>WPTabBarController+Swift</code>:</summary>

```swift
// WPTabBarController+Swift.swift

    func showBloggingPromptsFeatureIntroduction() {
//        if FeatureFlag.bloggingPrompts.enabled {
//            BloggingPromptsIntroductionPresenter().present(from: selectedViewController ?? self)
//        }
    }
```
</details>

### Testing prompt notifications

- Ensure that the Blogging Prompts flag is enabled, and the test code is applied.
- Tap on the Skip prompts context menu item.
- If you haven't authorized push for the first time, a system dialog should appear. Please grant the push authorization.
- Go to home screen, and wait about 5 seconds. A local notification should appear.
- Tap when the notification appears.
- Verify that the app correctly brings you to the post creation screen.

### Testing static prompt notifications

- To test static prompt notifications, comment out the `promptIDKey` line from the test code in the `DashboardPromptsCardCell`.
- Repeat the testing steps in the previous section.
- Verify that the notification shown is today's prompts.

Note: there's a known issue regarding prompt refresh in the dashboard causing outdated prompt, detailed here: https://github.com/wordpress-mobile/WordPress-iOS/issues/18626

### Tapping prompt notifications from different site

- Tap on the Skip prompts button.
- Quickly switch to another site, and go to home screen.
- Tap when the notification appears.
- Verify that the post creation flow creates the post for the first blog (where the skip prompts button is tapped).

## Regression Notes
1. Potential unintended areas of impact
N/A. Feature is unreleased.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A. Feature is unreleased.

3. What automated tests I added (or what prevented me from doing so)
N/A. Feature is unreleased.

PR submission checklist:
